### PR TITLE
Llmblog2

### PR DESCRIPTION
**Pull request by KevinMusgrave**

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
The `test_model.py` script downloads the best checkpoint if `exp_id` is given, and this was something I was able to run on my laptop for TinyLlama. I also ran it for the pretrained Mistral model, but I don't want to run it for a finetuned Mistral model because I think the checkpoint will be huge.

Part of the problem is that the checkpoint contains the optimizer state, which I don't need to evaluate the model. Maybe there is some way to download only part of the checkpoint? I'm not sure.

File: blog/llm-finetuning-2/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Finetuning Mistral-7B using LoRA and DeepSpeed
+
+In this demo, we finetune the [Mistral-7B](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2) using [LoRA](https://arxiv.org/abs/2106.09685) and [DeepSpeed](https://github.com/microsoft/DeepSpeed). We ran LoRA on two 80 GB A100 GPUs, and DeepSpeed on two, four, and eight 80 GB A100 GPUs.
```
Comment (aciborowska) [inline blog/llm-finetuning-2/README.md]:
```suggestion
In this demo, we finetune [Mistral-7B](https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.2) using [LoRA](https://arxiv.org/abs/2106.09685) and [DeepSpeed](https://github.com/microsoft/DeepSpeed). We ran LoRA on two 80 GB A100 GPUs, and DeepSpeed on two, four, and eight 80 GB A100 GPUs.
```

File: blog/llm-finetuning-2/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+To get started, first install Determined on your local machine:
+```bash
+pip install determined
+```
+
+Finetune with LoRA:
+```bash
+det e create distributed.yaml .
```
Comment (aciborowska) [inline blog/llm-finetuning-2/README.md]: should we just call it lora.yaml?

File: blog/llm-finetuning-2/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+## Configuration
+
+Change configuration options in `distributed.yaml`. Some important options are:
+- `slots_per_trial`: the number of GPUs to use.
+- `dataset_subset`: the difficulty subset to train on.
+- `per_device_train_batch_size`: the batch size per GPU.
+
+
+DeepSpeed configuration options are in the `ds_configs` folder.
```
Comment (aciborowska) [inline blog/llm-finetuning-2/README.md]:
```suggestion
DeepSpeed configuration files are in the `ds_configs` folder.
```

File: blog/llm-finetuning-2/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+## Testing
+
+Test your model's generation capabilities:
+
+```bash
+python test_model.py --exp_id <exp_id> --dataset_subset <dataset_subset>
```
Comment (aciborowska) [inline blog/llm-finetuning-2/README.md]: have you confirmed it still works with lora and deepspeed models?

File: blog/llm-finetuning-2/deepspeed.yaml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+name: mistral deepspeed easy
+debug: false
+environment:
+  environment_variables:
+    - NCCL_DEBUG=INFO
+  image: determinedai/genai-train:latest
```
Comment (aciborowska) [inline blog/llm-finetuning-2/deepspeed.yaml]: let's use determined image here. this one contains genai codebase. we will need to add deepspeed to the requirements.

File: blog/llm-finetuning-2/distributed.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
+name: mistral lora easy
+debug: false
+environment:
+  environment_variables:
+    - NCCL_DEBUG=INFO
+  image: determinedai/environments-dev:python-3.10-pytorch-2.0-deepspeed-0.10.0-smartsim
```
Comment (aciborowska) [inline blog/llm-finetuning-2/distributed.yaml]: let's use a regular det image here as well

File: blog/llm-finetuning-2/finetune.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        labels[mask] = tokenizer.pad_token_id
+        preds[mask] = tokenizer.pad_token_id
+
+        decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
+        decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)
+
+        for l, p in zip(decoded_labels, decoded_preds):
+            if l != p:
+                logging.error(f"decoded_label:{l}")
+                logging.error(f"decoded_pred:{p}")
```
Comment (aciborowska) [inline blog/llm-finetuning-2/finetune.py:99]: let's also remove this dummy logging

File: blog/llm-finetuning-2/finetune.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+            task_type="CAUSAL_LM",
+            inference_mode=False,
+            r=8,
+            lora_alpha=32,
+            lora_dropout=0.1,
+        )
+
+        model = get_peft_model(model, peft_config)
+
+    logging.error(f"dataset={dataset['train'][0]}")
```
Comment (aciborowska) [inline blog/llm-finetuning-2/finetune.py]:
```suggestion
```

File: blog/llm-finetuning-2/finetune.py (hunk 7 above)
Comment (aciborowska) [inline blog/llm-finetuning-2/finetune.py]:
```suggestion
```

File: blog/llm-finetuning-2/finetune.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    )
+
+    trainer.add_callback(det_callback)
+    # we need to comment this one out, since it will lead to the following error:
+    # [parameter_offload.py:86:_apply_to_tensors_only] A module has unknown inputs or outputs type (<class 'transformers.cache_utils.DynamicCache'>)
+    # and the tensors embedded in it cannot be detected. The ZeRO-3 hooks designed to trigger before or after backward pass of the module relies on
+    # knowing the input and output tensors and therefore may not get triggered properly.
+    # The error happens due to deepspeed initialization happening in the trainer.train(), hence call on eval fails.
+
+    # trainer.evaluate()
```
Comment (aciborowska) [inline blog/llm-finetuning-2/finetune.py]:
```suggestion
```

File: blog/llm-finetuning-2/README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+Test your model's generation capabilities:
+
+```bash
+python inference.py --exp_id <exp_id> --dataset_subset <dataset_subset>
+```
+
+Where 
+- `<exp_id>` is the id of your finetuning experiment in the Determined UI.
+- `<dataset_subset>` is one of "easy", "medium", or "hard".
```
Comment (aciborowska) [inline blog/llm-finetuning-2/README.md:43]:
```suggestion
- `<dataset_subset>` is one of "easy", "medium", or "hard".
- `<lora>` is a flag to indicate whether the model was finetuned with LoRA.
- `<device>` is one of "cuda" or "cpu".
```

**Review by aciborowska — APPROVED** [review]
LGTM!

**Pull request merged**